File: 2022-2023/sourcecode/lab/lab5/CarsyAppJava/app/src/main/java/pl/udu/uwr/pum/carsyappjava/adapters/OverviewCostsAdapter.java
package pl.udu.uwr.pum.carsyappjava.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.stream.Collectors;

import pl.udu.uwr.pum.carsyappjava.R;
import pl.udu.uwr.pum.carsyappjava.data.Car;
import pl.udu.uwr.pum.carsyappjava.data.Cost;
import pl.udu.uwr.pum.carsyappjava.data.CostType;
import pl.udu.uwr.pum.carsyappjava.data.DataProvider;
import pl.udu.uwr.pum.carsyappjava.util.FormatterUtil;

public class OverviewCostsAdapter extends RecyclerView.Adapter<OverviewCostsAdapter.ViewHolder> {

    private final ArrayList<Car> cars = DataProvider.getCars();

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView carNameTextView;
        private final TextView refuelingTextView;
        private final TextView serviceTextView;
        private final TextView parkingTextView;
        private final TextView insuranceTextView;
        private final TextView ticketTextView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            carNameTextView = itemView.findViewById(R.id.overviewRecyclerViewCarName);
            refuelingTextView = itemView.findViewById(R.id.overviewRecyclerViewFueling);
            serviceTextView = itemView.findViewById(R.id.overviewRecyclerViewService);
            parkingTextView = itemView.findViewById(R.id.overviewRecyclerViewParking);
            insuranceTextView = itemView.findViewById(R.id.overviewRecyclerViewInsurance);
            ticketTextView = itemView.findViewById(R.id.overviewRecyclerViewTicket);
        }

        public void bind (Car item){
            carNameTextView.setText(item.getName());
            refuelingTextView.setText(costValue(item, CostType.REFUELING));
            serviceTextView.setText(costValue(item, CostType.SERVICE));
            parkingTextView.setText(costValue(item, CostType.PARKING));
            insuranceTextView.setText(costValue(item, CostType.INSURANCE));
            ticketTextView.setText(costValue(item, CostType.TICKET));
        }

        private String costValue(Car item, CostType costType){
            return FormatterUtil.decimalFormat.format(
                    item.getCosts().stream()
                            .filter(i-> i.getType() == costType)
                            .map(Cost::getAmount)
                            .reduce(0, Integer::sum)
            ) + " zł";
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(
                R.layout.item_recyclerview_cost_overview, parent, false
        ));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Car item = cars.get(position);
        holder.bind(item);
    }

    @Override
    public int getItemCount() {
        return cars.size();
    }
}

File: 2022-2023/sourcecode/lab/lab5/CarsyAppKotlin/app/src/main/java/pl/udu/uwr/pum/carsyappkotlin/adapters/OverviewCostsAdapter.kt
package pl.udu.uwr.pum.carsyappkotlin.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import pl.udu.uwr.pum.carsyappkotlin.R
import pl.udu.uwr.pum.carsyappkotlin.data.Car
import pl.udu.uwr.pum.carsyappkotlin.data.Cost
import pl.udu.uwr.pum.carsyappkotlin.data.CostType
import pl.udu.uwr.pum.carsyappkotlin.data.DataProvider
import pl.udu.uwr.pum.carsyappkotlin.util.decimalFormat

class OverviewCostsAdapter : RecyclerView.Adapter<OverviewCostsAdapter.ViewHolder>() {

    private val cars = DataProvider.cars

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val carNameTextView: TextView = view.findViewById(R.id.overviewRecyclerViewCarName)
        private val refuelingTextView: TextView = view.findViewById(R.id.overviewRecyclerViewFueling)
        private val serviceTextView: TextView = view.findViewById(R.id.overviewRecyclerViewService)
        private val parkingTextView: TextView = view.findViewById(R.id.overviewRecyclerViewParking)
        private val insuranceTextView: TextView = view.findViewById(R.id.overviewRecyclerViewInsurance)
        private val ticketTextView: TextView = view.findViewById(R.id.overviewRecyclerViewTicket)

        fun bind (item: Car){
            carNameTextView.text = item.name
            refuelingTextView.text = costValue(item, CostType.REFUELING)
            serviceTextView.text = costValue(item, CostType.SERVICE)
            parkingTextView.text = costValue(item, CostType.PARKING)
            insuranceTextView.text = costValue(item, CostType.INSURANCE)
            ticketTextView.text = costValue(item, CostType.TICKET)
        }

        private fun costValue(item: Car, costType: CostType): String{
            return decimalFormat.format(item.costs
                .filter { it.type == costType }
                .sumOf { it.amount }).toString() + " zł"
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(
            R.layout.item_recyclerview_cost_overview, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = cars[position]
        holder.bind(item)
    }

    override fun getItemCount(): Int = cars.size
}